src/ser/serialize.rs

```diff
@@ -96,11 +99,20 @@ pub fn publish_message<'a, 'b, 'c>(
     // Write the variable header into the packet.
     packet.write_properties(properties)?;
 
-    // TODO: Handle packet ID when QoS > 0.
+    if qos != QoS::AtMostOnce {
+        packet.write_u16(id)?;
+    }
+
     packet.write_utf8_string(topic)?;
 
+    let flags :u8 = match qos {
+        QoS::AtMostOnce => 0,
+        QoS::AtLeastOnce => 2,
+        QoS::ExactlyOnce => 6,
+    };
```

Comment:
I would prefer we assigned QoS to be integral in value and then shift it over by 1 bit to follow the diagram for the fixed header. E.g.
```rust
use bit_field::BitField
let flags = 0u8.set_bit(1..=2, qos as u8)
                       .set_bit(3, duplicate)
```

or similar, as it then makes it less error prone and clear what's going on for flag settings

Reply:
I used your code for qos, but I do not want to use duplicate in this method. Instead I set it in session_state when it is saved because any retries need the DUP value set but at that point this serialize method is not called.

I hope I'm not forgetting anything else that can potentially change between resends.

```rust
        let mut buf: Vec<u8, MSG_SIZE> = Vec::from_slice(packet).unwrap();
        // Set DUP = 1 (bit 3). If this packet is ever read it's just because we want to resend it
        buf[0] |= 1 << 3;